### DISCUSSION TITLE
Seeing IPsec decrypted traffic on the ipsec ingress interface as outbound traffic

### DISCUSSION THREAD

**Discussion by cyanideangel2000** [strongswan/strongswan#1444] [Q&A] [answered]
Hello Team,

I have a simple setup running ubuntu 20.04 strongswan 5.9.7 as the ipsec responder only.
We rely on netfilter packet mark to route the traffic to a tuntap interface post xfrm decryption.
Everything works fine. Ipsec comes in, gets xfrm decrypted, iptables and ip rules do their thing and eventually packet hits the tuntap interface.
However, while taking a tcpdump on the physical interface where ipsec 4500 traffic is coming into the box, i noticed I am seeing the inner packet in cleartext as the outbound traffic.
I ran the following
```tcpdump -nni eth1 port 443 and outbound```
This shows me the cleartext traffic which threw me off a bit. For sure on the tuntap i see the traffic as well.
Now as per this: https://upload.wikimedia.org/wikipedia/commons/3/37/Netfilter-packet-flow.svg
Post decode, traffic should re-enter in raw prerouting chain not the interface.

Further I don't see all the packets there. On tuntap i see every packet, but on the physical i see a subset of them.
Any suggestions what might be going on here. I am not able to explain why that would be the case.

**Comment by tobiasbrunner** [comment, marked as answer]
> However, while taking a tcpdump on the physical interface where ipsec 4500 traffic is coming into the box, i noticed I am seeing the inner packet in cleartext as the outbound traffic.

Inbound plaintext packets are always seen on the interface on which the ESP packet was received. If you are referring to the `outbound` filter that might not work accurately for such packets are they have not proper link-layer header. But you could try reporting that to the netdev kernel mailing list, maybe it's something that could be corrected.

Also note that he diagram you refer to is not accurate or complete by any means in regards to IPsec.

> Further I don't see all the packets there.

What packets don't you see? If you are referring to actual outbound traffic, that can't be captured via libpcap, see [this FAQ entry](https://docs.strongswan.org/docs/5.9/support/faq.html#_capturing_outbound_plaintext_packets_with_tcpdumpwireshark).

**Reply by cpinkham-lumn** [reply]
@tobiasbrunner can you reference a more complete flow diagram that applies where IPSEC is being used?  The diagram linked above is the one referenced on the FAQ page you linked to. 

I am troubleshooting an issue where SYN ACK reply packets coming back through the VPN are decrypted, but are not flowing through the prerouting chains or connection tracking.  The initial SYN packet goes through a iptables SNAT rule before going into the VPN on its way out to the remote server, but the SYN ACK reply is jumping straight to the mangle input chain after decryption and is not having its destination IP changed to match the original source IP from before the SNAT.  Due to this, the system sees a SYN ACK packet coming to a port which is not valid on its local IP and sends a RST packet back to the remote server over the VPN.  The VPN is working fine when packets are sourced from the local IP because no address translation is involved on the reply side, but when address translation is required, the reply traffic address translation is not working.  I am trying to track down whether this may be due to using a bridge interface instead of having the IP on a physical interface, but the flow diagram linked above and linked from the FAQ does not appear to be correct for what I am seeing when tracing the packets using -j TRACE.

**Reply by tobiasbrunner** [reply]
I don't know any better diagram, sorry. But just to see if TCP generally works via SNAT before IPsec, I added an SSH command from alice to bob to the [ikev2/nat-virtual-ip test scenario](https://strongswan.org/testing/testresults/ikev2/nat-virtual-ip/). This worked fine (except for an MTU issue), so this might be due to your network/firewall config or the kernel you're using.

**Reply by cpinkham-lumn** [reply]
Thanks, I'll keep digging and test without the bridge.  In my case TCP works if it doesn't have to go through the SNAT. Neither simple ping/icmp or TCP work with the SNAT since the reply doesn't have it's destination IP changed.  The source of that original request requiring SNAT could be a remote system on the local LAN routing through the box or a VM or container on the local box, neither works if the SNAT is required.  If I find something I'll post here just for reference.

**Comment by cpinkham-lumn** [comment]
My issue seems to be bridge related.  I put the IP directly on a NIC and the SNAT through VPN works fine and I can see the decrypted traffic jumping back to the prerouting chains while it does not do that when using a bridge interface for the external interface.  This may be related to the same non-strongswan issue described in https://github.com/strongswan/strongswan/discussions/1450 as I was also seeing non-established connections in conntrack because the reply packets were skipping prerouting and connection tracking.  Thanks for your feedback.
